src/psyclone/nemo.py

```diff
@@ -0,0 +1,822 @@
+        Also overrides the PSy gen method so that we generate GOcean-
+        specific PSy module code. '''
+
+    def __init__(self, ast):
+
+        self._name = "NEMO-PSY"  # TODO use a meaningful name
+        self._invokes = NemoInvokes(ast)
+        self._ast = ast
+
+    def inline(self, module):
+        # Override base-class method because we don't yet support it
+        raise NotImplementedError("The NemoPSy.inline method has not yet "
+                                  "been implemented!")
+
+    @property
+    def gen(self):
+        '''
+        Generate the (updated) fparser2 AST for the NEMO code represented
+        by this NemoPSy object.
+
+        :rtype: :py:class:`fparser.two.Fortran2003.Main_Program` or \
+                :py:class:`fparser.two.Fortran2003.Subroutine_Subprogram` or \
+                :py:class:`fparser.two.Fortran2003.Function_Subprogram`.
+        '''
+        # Walk down our Schedule and update the underlying fparser2 AST
+        # to account for any transformations
+        self.invokes.update()
+
+        # Return the fparser2 AST
+        return self._ast
+
+
+class NemoSchedule(Schedule, ASTProcessor):
+    '''
+    The NEMO-specific schedule class. This is the top-level node in
+    PSyclone's representation of a NEMO program unit (program,
+    subroutine etc).
+
+    :param invoke: The Invoke to which this Schedule belongs.
+    :type invoke: :py:class:`psyclone.nemo.NemoInvoke`
+    :param ast: the fparser2 AST of the NEMO code for which to generate \
+                a Schedule.
+    :type ast: :py:class:`fparser.two.Fortran2003.Main_Program` or \
+               :py:class:`fparser.two.Fortran2003.Subroutine_Subprogram` or \
+               :py:class:`fparser.two.Fortran2003.Function_Subprogram`.
+    '''
+
+    def __init__(self, invoke, ast):
+        Node.__init__(self)
+
+        self._invoke = invoke
+        self._ast = ast
+        self.process_nodes(self, ast.content, ast)
+
+
+    def view(self, indent=0):
+        ''' Print a representation of this NemoSchedule '''
+        print(self.indent(indent) + self.coloured_text + "[]")
+        for entity in self._children:
+            entity.view(indent=indent + 1)
+
+    def __str__(self):
+        ''' Returns the string representation of this NemoSchedule '''
+        result = "NemoSchedule():\n"
+        for entity in self._children:
+            result += str(entity)+"\n"
+        result += "End Schedule"
+        return result
+
+
+class NemoCodeBlock(Node):
+    ''' Node representing some generic Fortran code that PSyclone
+    does not attempt to manipulate. As such it is a leaf in the PSyclone
+    AST and therefore has no children. '''
+
+    def __init__(self, statements, parent=None):
+        Node.__init__(self, parent=parent)
+        # Store a list of the parser objects holding the code
+        # associated with this block
+        self._statements = statements[:]
```

Comment:
The contents of the list need to be copied because the list itself is temporary and is deleted (in the renamed `add_code_block()` routine). Possibly this indicates something sub-optimal in the way I've implemented code block generation but I can't quite see what it is.